examples/h_mountain.py

```diff
@@ -0,0 +1,179 @@
+from gusto import *
+from firedrake import FunctionSpace, as_vector, \
+    VectorFunctionSpace, PeriodicIntervalMesh, ExtrudedMesh, \
+    SpatialCoordinate, exp, pi, cos, Function, conditional, Mesh, sin, op2, sqrt
+import sys
+
+dt = 5.0
+if '--running-tests' in sys.argv:
+    tmax = dt
+else:
+    tmax = 15000.
+
+res = 10
+nlayers = res*20 # horizontal layers
+columns = res*12 # number of columns
+L = 240000.
+m = PeriodicIntervalMesh(columns, L)
+
+# build volume mesh
+H = 50000. # Height position of the model top
+ext_mesh = ExtrudedMesh(m, layers=nlayers, layer_height=H/nlayers)
+Vc = VectorFunctionSpace(ext_mesh, "DG", 2)
+coord = SpatialCoordinate(ext_mesh)
+x = Function(Vc).interpolate(as_vector([coord[0], coord[1]]))
+a = 10000.
+xc = L/2.
+x, z = SpatialCoordinate(ext_mesh)
+hm = 1.
+zs = hm*a**2/((x-xc)**2 + a**2)
+smooth_z = True
+if smooth_z:
+    zh = 5000.
+    xexpr = as_vector([x, conditional(z < zh, z + cos(0.5*pi*z/zh)**6*zs, z)])
+else:
+    xexpr = as_vector([x, z + ((H-z)/H)*zs])
+new_coords = Function(Vc).interpolate(xexpr)
+mesh = Mesh(new_coords)
+
+# sponge function
+W_DG = FunctionSpace(mesh, "DG", 2)
+x, z = SpatialCoordinate(mesh)
+zc = H-20000.
+mubar = 0.3/dt
+mu_top = conditional(z <= zc, 0.0, mubar*sin((pi/2.)*(z-zc)/(H-zc))**2)
+mu = Function(W_DG).interpolate(mu_top)
+fieldlist = ['u', 'rho', 'theta']
+timestepping = TimesteppingParameters(dt=dt, alpha=1.0)
+output = OutputParameters(dirname='h_mountain_smootherz', dumpfreq=30, dumplist=['u'], perturbation_fields=['theta', 'rho'])
+parameters = CompressibleParameters(g=9.80665, cp=1004.)
+diagnostics = Diagnostics(*fieldlist)
+diagnostic_fields = [CourantNumber(), VelocityZ(), HydrostaticImbalance()]
+
+state = State(mesh, vertical_degree=1, horizontal_degree=1,
+              family="CG",
+              sponge_function=mu,
+              hydrostatic=True,
+              timestepping=timestepping,
+              output=output,
+              parameters=parameters,
+              diagnostics=diagnostics,
+              fieldlist=fieldlist,
+              diagnostic_fields=diagnostic_fields)
+
+
+# Off-centring parameter is set to alpha=1.0 for first time step
+# to omit spurious orography waves, then alpha=0.5
+def implicit_first_step():
+    if state.t.dat.data_ro[0] == dt:
+        state.timestepping.alpha = 0.5
+
+state.parameter_update = implicit_first_step
```

Comment:
Rather than closing over `state` in the `parameter_update` method, you should change the API so that it is clear what `parameter_update` can change.  You could do this by making this method take an argument.

A better way to do this would be to pass the update function in as an argument to the `State` constructor.

Then you could have something like:
```
class State(...):

    def __init__(self, ..., parameter_update=None):
       ...
       if parameter_update is not None:
           self.parameter_update = parameter_update
       else:
           self.parameter_update = lambda state: pass

      ...
     def run(self, ...):
        while ...:
             self.parameter_update(self)
             ...
```

Reply:
Although having looked at this.  This update function is not a property of the `State`, but rather the time loop object.  So the update callback should live there.